File: backend/app/protocol.py
"""Protocol models derived from protocol_v1.md."""
from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, Field

from app.config import settings


# === Enums from protocol_v1.md ===


class SpinMode(str, Enum):
    """Spin mode from protocol_v1.md."""

    NORMAL = "NORMAL"
    BUY_FEATURE = "BUY_FEATURE"


class GameMode(str, Enum):
    """Game mode for nextState."""

    BASE = "BASE"
    FREE_SPINS = "FREE_SPINS"


class EventType(str, Enum):
    """Event types from protocol_v1.md section 4."""

    REVEAL = "reveal"
    WIN_LINE = "winLine"
    ENTER_FREE_SPINS = "enterFreeSpins"
    HEAT_UPDATE = "heatUpdate"
    EVENT_START = "eventStart"
    EVENT_END = "eventEnd"
    SPOTLIGHT_WILDS = "spotlightWilds"
    WIN_TIER = "winTier"
    BONUS_END = "bonusEnd"


class GameEventType(str, Enum):
    """eventType values for eventStart/eventEnd."""

    BOOST = "boost"
    RAGE = "rage"
    EXPLOSIVE = "explosive"
    BONUS = "bonus"
    FINALE = "finale"


class WinTier(str, Enum):
    """Win tier from protocol_v1.md."""

    NONE = "none"
    BIG = "big"
    MEGA = "mega"
    EPIC = "epic"


class FinalePath(str, Enum):
    """Finale path from protocol_v1.md."""

    UPGRADE = "upgrade"
    MULTIPLIER = "multiplier"
    STANDARD = "standard"


# === Request Models ===


class SpinRequest(BaseModel):
    """POST /spin request body from protocol_v1.md."""

    clientRequestId: str = Field(..., description="UUIDv4 idempotency key")
    betAmount: float = Field(..., description="Must be in allowedBets")
    mode: SpinMode = Field(default=SpinMode.NORMAL)
    hypeMode: bool = Field(default=False)


# === Response Models ===


class Configuration(BaseModel):
    """Configuration object in /init response."""

    currency: str = "USD"
    allowedBets: list[float] = settings.allowed_bets
    enableBuyFeature: bool = settings.enable_buy_feature
    buyFeatureCostMultiplier: int = settings.buy_feature_cost_multiplier
    enableTurbo: bool = settings.enable_turbo
    enableHypeModeAnteBet: bool = settings.enable_hype_mode_ante_bet
    hypeModeCostIncrease: float = settings.hype_mode_cost_increase


class RestoreState(BaseModel):
    """Restore state for unfinished rounds."""

    mode: GameMode
    spinsRemaining: int
    heatLevel: int


class InitResponse(BaseModel):
    """GET /init response from protocol_v1.md."""

    protocolVersion: str = settings.protocol_version
    configuration: Configuration = Field(default_factory=Configuration)
    restoreState: RestoreState | None = None


class Context(BaseModel):
    """Context object in spin response."""

    currency: str = "USD"


class Outcome(BaseModel):
    """Outcome object in spin response."""

    totalWin: float
    totalWinX: float
    isCapped: bool = False
    capReason: str | None = None


class NextState(BaseModel):
    """Next state after spin."""

    mode: GameMode = GameMode.BASE
    spinsRemaining: int = 0
    heatLevel: int = 0


class SpinResponse(BaseModel):
    """POST /spin response from protocol_v1.md."""

    protocolVersion: str = settings.protocol_version
    roundId: str
    context: Context = Field(default_factory=Context)
    outcome: Outcome
    events: list[dict[str, Any]] = Field(default_factory=list)
    nextState: NextState = Field(default_factory=NextState)

File: frontend/src/types/protocol.ts
/**
 * Protocol types derived from protocol_v1.md
 * Source of truth: /protocol_v1.md
 */

/** Configuration returned from /init */
export interface Configuration {
  currency: string
  allowedBets: number[]
  enableBuyFeature: boolean
  buyFeatureCostMultiplier: number
  enableTurbo: boolean
  enableHypeModeAnteBet: boolean
  hypeModeCostIncrease: number
}

/** Restore state if player has unfinished free-spins */
export interface RestoreState {
  mode: 'FREE_SPINS'
  spinsRemaining: number
  heatLevel: number
}

/** Response from GET /init */
export interface InitResponse {
  protocolVersion: string
  configuration: Configuration
  restoreState: RestoreState | null
}

/** Request body for POST /spin */
export interface SpinRequest {
  clientRequestId: string
  betAmount: number
  mode: 'NORMAL' | 'BUY_FEATURE'
  hypeMode: boolean
}

/** Context included in spin response */
export interface SpinContext {
  currency: string
}

/** Outcome of a spin */
export interface Outcome {
  totalWin: number
  totalWinX: number
  isCapped: boolean
  capReason: 'max_win_base' | 'max_win_bonus' | 'max_exposure' | null
}

/** Next state after spin */
export interface NextState {
  mode: 'BASE' | 'FREE_SPINS'
  spinsRemaining: number
  heatLevel: number
}

/** Game event - see events.ts for full union type */
import type { GameEvent } from './events'

/** Response from POST /spin */
export interface SpinResponse {
  protocolVersion: string
  roundId: string
  context: SpinContext
  outcome: Outcome
  events: GameEvent[]
  nextState: NextState
}

/** Error object structure */
export interface ErrorInfo {
  code: ErrorCode
  message: string
  recoverable: boolean
}

/** Error response from server */
export interface ErrorResponse {
  protocolVersion: string
  error: ErrorInfo
}

/** Error codes from error_codes.md */
export type ErrorCode =
  | 'INVALID_REQUEST'
  | 'INVALID_BET'
  | 'FEATURE_DISABLED'
  | 'INSUFFICIENT_FUNDS'
  | 'ROUND_IN_PROGRESS'
  | 'IDEMPOTENCY_CONFLICT'
  | 'RATE_LIMIT_EXCEEDED'
  | 'MAINTENANCE'
  | 'INTERNAL_ERROR'
  | 'NOT_IMPLEMENTED'
